### PR TITLE
Add metric tags to desired lrp update

### PR DESCRIPTION
**Pull request by acrmp**
## Please make sure to complete all of the following steps

1. Check the [Contributing document](https://github.com/cloudfoundry/diego-release/blob/develop/CONTRIBUTING.md) on how to sign the CLA and run tests.
1. Submit your PR to this repo.
1. [**Submit an accompanying PR Review Request**](https://github.com/cloudfoundry/diego-release/issues/new?assignees=&labels=&template=pr-review-request.md&title=%5BBBS+PR+REVIEW%5D%3A) referencing this PR so the Diego Team knows to review your pull request. 
* **Note: this PR will not be reviewed unless you submit the [PR Review Request](https://github.com/cloudfoundry/diego-release/issues/new?assignees=&labels=&template=pr-review-request.md&title=%5BBS+BPR+REVIEW%5D%3A)**.

***************************

### What is this change about?

* Extend DesiredLRPUpdate to allow clients to provide metric tags to be updated
* As part of this we bumped the revision of the `/v1/desired_lrp/update` endpoint
* Update documentation to reflect the changes to the endpoint
* We changed how counterfeiter is invoked

We are submitting this PR against the `update-metric-tags` branch. We intend to submit further PRs against that branch for y'all to review individually. We don't expect these changes to be merged into `main` until you have reviewed all the PRs related to these changes.

### What problem it is trying to solve?

Clients want to be able to dynamically modify metric tags, for example an application name.

Subsequent PRs will follow to actually have the logs and metrics emitted use the updated tags.

### What is the impact if the change is not made?

Clients are not able to update metric tags without creating a new DesiredLRP.

### How should this change be described in diego-release release notes?

Extend DesiredLRPUpdate to allow clients to provide metric tags to be updated

### Please provide any contextual information.

https://github.com/cloudfoundry/diego-release/issues/662

#### CreatedAt

* We persist the updated metric tags to the database by updating the RunInfo. To create a RunInfo we needed to pass a CreatedAt timestamp. We used the current clock time which we *think* is ok.
* We saw that CreatedAt was added to the RunInfo in this commit aaea01fb94c38b60ef62461e9b5b4894a263bb9a.
* We think that it was used to remove orphaned RunInfo's and we're not sure if it's being used following the removal of etcd code in: 09fd338b2a8c0d19040eae0f1af280216a9219b6?

#### Documentation

* As part of this change we updated docs for the Desired LRP update endpoint.
* We saw [a reference to n-sync in this doc](https://github.com/cloudfoundry/bbs/blob/main/doc/lrps.md#the-lrp-lifecycle) which we think may be out of date.

#### Tests

* We've run the unit tests via the script that wraps docker (`./scripts/docker-test`).
* We've not run other diego test suites.
* We've tried to be conscious of upgrade scenarios but have not done a lot of testing around those at this point.

### Tag your pair, your PM, and/or team!

@rroberts2222 @ctlong

Thank you!

### CODE REVIEW

File: models/desired_lrp.go [hunk 1]
```diff
@@ -286,10 +286,6 @@ func (d *DesiredLRP) Copy() *DesiredLRP {
 	return &newDesired
 }
 
-func (d *DesiredLRP) CreateComponents(createdAt time.Time) (DesiredLRPSchedulingInfo, DesiredLRPRunInfo) {
-	return d.DesiredLRPSchedulingInfo(), d.DesiredLRPRunInfo(createdAt)
-}
-
```
Comment (geofffranks) [inline models/desired_lrp.go:292]: Why was this function removed?

**Comment by geofffranks** [comment]
Added one specific question, but also have a larger/general question about how this API should behave, which @acrmp brought up in Slack.

Is the intention of this to require the MetricsTag property on every UpdateDesiredLRP call ever made? If not, it might make more sense to add it as an optional property when validating, since it is an optional property when inserting into the database. At that point, I'm not sure bumping the API revision is necessary, but @aminjam might have a better opinion here. 

Hoever if the intention is that every UpdateDesiredLRP moving forward should fail if it doesn't include MetricsTag data, bumping the revision on the Update API seems like the right move.  We would want to ensure that anything calling UpdateDesiredLRP gets updated with new client code, and make a note about deprecating this version of the API in the future.

File: models/desired_lrp.go (hunk 1 above)
Comment (acrmp) [reply models/desired_lrp.go:292]: We couldn't see where this function was used and removed it in 6f0b40377470eb while we were here.

**Comment by acrmp** [comment]
@geofffranks Thanks for taking a look.

We don't intend for the metric tags to be required on every UpdateDesiredLRP call. We have some tests in the PR to ensure that Validate succeeds when metric tags not passed.

We also weren't sure if an API revision was entirely necessary. We were attempting to follow the [guidance in these docs](https://github.com/cloudfoundry/bbs/blob/c0a3fffd5a1c8c6c96c0ca49057fd53024ddcd7b/doc/revisioning-bbs-api-endpoints.md#api-endpoint-and-handler-naming-conventions) which suggest that an API revision is necessary even when adding a new field.

**Comment by acrmp** [comment]
@geofffranks @aminjam We took a look at the history of revision changes to the `/v1/desired_lrp/desire` endpoint for comparison.

We saw that for an additive change (for example the introduction of sidecars f933be976e4bdae4e54915a8) it seems that we did not introduce a new revision for the endpoint.

For structural changes (for example changes to how volume mounts 6a1edd0d72ddf279c2 or time f59ac010a939b was represented) we saw that the endpoint revision was bumped.

We think our conclusion is that for this change the endpoint does not require a new revision, so based on that we'll remove that from the PR. Does that sound correct to you?

**Comment by acrmp** [comment]
@geofffranks @winkingturtle-vmw We pushed another commit 60aa2a4a02f39a71c06f to revert the revision bump as agreed in slack.

**Review by ebroberson — COMMENTED** [review]
LGTM but want another approver

**Review by mariash — APPROVED** [review]

**Comment by mariash** [comment]
@acrmp we think PR looks good. Why do you think it should be merged into a feature branch? Maybe let's just keep this PR open to keep all the context around.

**Comment by ctlong** [comment]
@mariash merging into a feature branch was the approach that we had previously agreed on with @ameowlia. We've had success with in the past for larger tracks of work that can't be broken up, largely because it allows us to make granular PRs for easier review without having to worry too much about merge conflicts.

❓ Does that sound okay to you? Happy to chat more about this in Slack.

**Comment by mariash** [comment]
Sounds good to me

**Pull request merged**